fluent/fluent.go

```diff
@@ -282,31 +289,71 @@ func e(x, y float64) int {
 }
 
 func (f *Fluent) reconnect() {
-	for i := 0; ; i++ {
-		err := f.connect()
+	var err error
+	for i := 0; i < f.Config.MaxRetry; i++ {
+		err = f.connect()
 		if err == nil {
-			f.send()
-			return
-		}
-		if i == f.Config.MaxRetry {
-			// TODO: What we can do when connection failed MaxRetry times?
-			panic("fluent#reconnect: failed to reconnect!")
+			break
 		}
 		waitTime := f.Config.RetryWait * e(defaultReconnectWaitIncreRate, float64(i-1))
 		time.Sleep(time.Duration(waitTime) * time.Millisecond)
 	}
+
+	if err != nil && !f.UseReconnectError {
+		panic("fluent#reconnect: failed to reconnect!")
+	}
+
+	f.muconn.Lock()
+	f.reconnectErr = err
+	f.muconn.Unlock()
+
+	if err == nil {
+		f.send()
+	}
+}
+
+// Reconnect re-initializes a connection.
+func (f *Fluent) Reconnect() error {
+	f.muconn.Lock()
+	defer f.muconn.Unlock()
+
+	for f.reconnecting {
+		f.connWait.Wait()
+	}
+
+	if f.conn != nil {
+		return nil
+	}
+
+	f.reconnecting = true
+	go f.reconnect()
```

Comment:
This looks to be called in parallel with `go f.reconnect()` in L354. Is it correct and intentional?